Introduce support for running code in fibers

daandemeyer opened the pull request:
Traditionally, asynchronous programming in systemd has been achieved using
sd-event along with the asynchronous interfaces of sd-bus and sd-varlink.
This works well when the system is reacting to events and all code triggered
by those events can run without blocking. In these scenarios, the global
Manager object is passed as userdata to the callback, and the callback can
use the stack as usual, declaring local state and ensuring proper cleanup via
_cleanup_. Control flow structures, such as loops, work as expected, and
everything runs smoothly.

However, challenges arise when the code needs to perform long-running
operations within these callbacks. Since the system cannot block execution
within the callback, we can't directly invoke a long-running operation and
wait for its result without introducing complexities. Instead, we need to
initiate the long-running task, register for completion with sd-event,
sd-bus, or sd-varlink, and provide a callback to be invoked when the
operation completes.

This callback, however, only receives a single userdata pointer, which
forces us to bundle all local variables into a struct and pass it along as
part of the callback. On top of that, after queuing the asynchronous
operation, the caller continues executing. As the caller's stack unwinds
when the function exits, the resources and state within the local scope may
be prematurely cleaned up. Therefore, the struct must store copies of the
local variables or ensure proper reference counting to prevent premature
resource cleanup.

When multiple long-running operations need to be initiated within a loop,
the complexity grows further. We must introduce additional shared state to
track the completion of all operations before we can run any code that
depends on their results.

Furthermore, since the daemon may be shut down at any time, we must track
the lifecycle of each long-running operation in the global Manager struct,
ensuring proper cleanup even when stack unwinding can no longer manage the
resources for us.

Fibers, or green threads, provide a more natural way of handling
asynchronous operations. By enabling cooperative multitasking within a
single thread, fibers allow us to write code that looks like it’s running
synchronously, but with the ability to yield control at predefined points,
such as when waiting for long-running tasks to complete.

With fibers, we can simplify the control flow by running asynchronous
operations within a fiber, allowing us to "pause" execution while waiting
for the long-running operation to finish and then "resume" the operation once
it's complete. This eliminates the need for multiple callback chains,
extensive state tracking, and the potential pitfalls of stack unwinding.

This commit introduces the ability to execute long-running operations in a
non-blocking manner while maintaining the simplicity and readability of
synchronous code. The fiber-based approach will significantly improve the
handling of complex workflows, making the code easier to write and maintain.

The implementation is based on ucontext.h's makecontext() (with a fallback
to the venerable sigaltstack() approach on musl), sigsetjmp()/siglongjmp() 
and sd-event. ucontext.h provides us with alternate stacks that we can switch 
between. We use sigsetjmp()/siglongjmp() instead of swapcontext() because the
latter forcibly saves/restores a per context signal mask every time it is called.
Using sigsetjmp()/siglongjmp(), we can avoid the unnecessary syscall and maintain
a per thread signal mask, which makes much more sense than having a per fiber
signal mask.

The default stack size is the same as a regular thread. Because we 
use mmap() to allocate the stack, the memory won't actually be used until it 
is paged in by the kernel, so we don't actually use 8MB per fiber.

To integrate fibers with the event loop, each fiber is assigned a deferred
event source which resumes the fiber when enabled. The deferred event source
is oneshot by default so the fiber will run immediately until it yields or
suspends. If it yields, the deferred event source is enabled again (oneshot)
immediately. If it suspends, before it suspends, one or more event sources
are registered with sd-event that will enable the deferred event source
(oneshot) to resume the fiber once the operation it is waiting for completes.

Yielding or suspending the fiber is done by calling sd_fiber_yield() or
sd_fiber_suspend() respectively. Both of these return zero on success or any
error value from the async operation that caused the fiber to resume.

This is also how fiber cancellation is implemented. When a fiber is cancelled,
sd_fiber_yield() and sd_fiber_suspend() will return ECANCELED when the fiber
is resumed, allowing the fiber to unwind its stack (which allows cleanup to
happen automatically) and finish.

Instead of having applications work directly with fibers, we hide them behind
a generic futures interface to represent long-running operations, regardless of
whether those operations are running on a fiber or not. Aside from fibers, the
futures library (sd-future) will for example allow waiting for sd-event sources 
and doing sd-bus calls in the background as well. Fibers can suspend until a 
future is ready with sd_fiber_await() or by having the future wake up the fiber
explicitly in its callback. A future always defaults to waking up the current 
fiber.

Each future kind plugs into the library by providing an sd_future_ops vtable
(alloc, free, cancel, set_priority). The library treats the impl pointer
returned by alloc() as a black box. Future Implementations retrieve it via 
sd_future_get_private().

A future starts in SD_FUTURE_PENDING and transitions exactly once to
SD_FUTURE_RESOLVED, carrying an integer result. Consumers can react to that
transition either by installing a one-shot callback with
sd_future_set_callback() (callback-style code) or by waiting on it from a
fiber via sd_fiber_await() (synchronous-looking fiber code). sd_fiber_await()
is itself built on a "wait future" that resolves when its target resolves;
sd_future_new_wait() exposes the same primitive directly so non-fiber callers
can chain futures without involving a fiber.

Cancellation is cooperative: sd_future_cancel() invokes the future impl's 
cancel callback, which is responsible for tearing down its work and ultimately
resolving the promise with -ECANCELED. For fiber futures this is what
surfaces as the ECANCELED return from sd_fiber_yield()/sd_fiber_suspend()
mentioned above.

Fire-and-forget fibers — created by passing a NULL ret to sd_fiber_new() —
take a self-reference on their future so they outlive the caller's scope.
The self-ref is dropped when the fiber resolves. This floating mechanism
(sd_fiber_set_floating()) is restricted to fiber futures because they
uniquely guarantee resolution; allowing it for arbitrary future kinds would
risk silent leaks for kinds that may never resolve.

Note that fiber cleanup depends on the runtime operating normally. Each
fiber's _cleanup_-style cleanups live on the fiber's own stack and run
only when the fiber is resumed and allowed to unwind, which requires a
working event loop to drive it to completion. The exit event source
registered for top-level fibers ensures unwind on a normal sd_event_exit(),
but if the event loop itself terminates abnormally (e.g. an unrecoverable
allocation failure mid-dispatch) before all fibers have resolved, their
stacks never unwind and any resources they own leak.

The code lives in libsystemd as sd-future (not exported) for the following reasons:
- We may want to make this a public libsystemd API in the future
- The code can't live in src/basic as it makes heavy use of sd-event
- The code can't live in src/shared as sd-bus and sd-event make use of it

The log and log-context headers are updated with functions to allow
fibers to have their own log prefix and log context.

daandemeyer committed:
Introduce support for running code in fibers

Traditionally, asynchronous programming in systemd has been achieved using
sd-event along with the asynchronous interfaces of sd-bus and sd-varlink.
This works well when the system is reacting to events and all code triggered
by those events can run without blocking. In these scenarios, the global
Manager object is passed as userdata to the callback, and the callback can
use the stack as usual, declaring local state and ensuring proper cleanup via
_cleanup_. Control flow structures, such as loops, work as expected, and
everything runs smoothly.

However, challenges arise when the code needs to perform long-running
operations within these callbacks. Since the system cannot block execution
within the callback, we can't directly invoke a long-running operation and
wait for its result without introducing complexities. Instead, we need to
initiate the long-running task, register for completion with sd-event,
sd-bus, or sd-varlink, and provide a callback to be invoked when the
operation completes.

This callback, however, only receives a single userdata pointer, which
forces us to bundle all local variables into a struct and pass it along as
part of the callback. On top of that, after queuing the asynchronous
operation, the caller continues executing. As the caller's stack unwinds
when the function exits, the resources and state within the local scope may
be prematurely cleaned up. Therefore, the struct must store copies of the
local variables or ensure proper reference counting to prevent premature
resource cleanup.

When multiple long-running operations need to be initiated within a loop,
the complexity grows further. We must introduce additional shared state to
track the completion of all operations before we can run any code that
depends on their results.

Furthermore, since the daemon may be shut down at any time, we must track
the lifecycle of each long-running operation in the global Manager struct,
ensuring proper cleanup even when stack unwinding can no longer manage the
resources for us.

Fibers, or green threads, provide a more natural way of handling
asynchronous operations. By enabling cooperative multitasking within a
single thread, fibers allow us to write code that looks like it’s running
synchronously, but with the ability to yield control at predefined points,
such as when waiting for long-running tasks to complete.

With fibers, we can simplify the control flow by running asynchronous
operations within a fiber, allowing us to "pause" execution while waiting
for the long-running operation to finish and then "resume" the operation once
it's complete. This eliminates the need for multiple callback chains,
extensive state tracking, and the potential pitfalls of stack unwinding.

This commit introduces the ability to execute long-running operations in a
non-blocking manner while maintaining the simplicity and readability of
synchronous code. The fiber-based approach will significantly improve the
handling of complex workflows, making the code easier to write and maintain.

The implementation is based on ucontext.h's makecontext() (with a fallback
to the venerable sigaltstack() approach on musl), sigsetjmp()/siglongjmp() 
and sd-event. ucontext.h provides us with alternate stacks that we can switch 
between. We use sigsetjmp()/siglongjmp() instead of swapcontext() because the
latter forcibly saves/restores a per context signal mask every time it is called.
Using sigsetjmp()/siglongjmp(), we can avoid the unnecessary syscall and maintain
a per thread signal mask, which makes much more sense than having a per fiber
signal mask.

The default stack size is the same as a regular thread. Because we 
use mmap() to allocate the stack, the memory won't actually be used until it 
is paged in by the kernel, so we don't actually use 8MB per fiber.

To integrate fibers with the event loop, each fiber is assigned a deferred
event source which resumes the fiber when enabled. The deferred event source
is oneshot by default so the fiber will run immediately until it yields or
suspends. If it yields, the deferred event source is enabled again (oneshot)
immediately. If it suspends, before it suspends, one or more event sources
are registered with sd-event that will enable the deferred event source
(oneshot) to resume the fiber once the operation it is waiting for completes.

Yielding or suspending the fiber is done by calling sd_fiber_yield() or
sd_fiber_suspend() respectively. Both of these return zero on success or any
error value from the async operation that caused the fiber to resume.

This is also how fiber cancellation is implemented. When a fiber is cancelled,
sd_fiber_yield() and sd_fiber_suspend() will return ECANCELED when the fiber
is resumed, allowing the fiber to unwind its stack (which allows cleanup to
happen automatically) and finish.

Instead of having applications work directly with fibers, we hide them behind
a generic futures interface to represent long-running operations, regardless of
whether those operations are running on a fiber or not. Aside from fibers, the
futures library (sd-future) will for example allow waiting for sd-event sources 
and doing sd-bus calls in the background as well. Fibers can suspend until a 
future is ready with sd_fiber_await() or by having the future wake up the fiber
explicitly in its callback. A future always defaults to waking up the current 
fiber.

Each future kind plugs into the library by providing an sd_future_ops vtable
(alloc, free, cancel, set_priority). The library treats the impl pointer
returned by alloc() as a black box. Future Implementations retrieve it via 
sd_future_get_private().

A future starts in SD_FUTURE_PENDING and transitions exactly once to
SD_FUTURE_RESOLVED, carrying an integer result. Consumers can react to that
transition either by installing a one-shot callback with
sd_future_set_callback() (callback-style code) or by waiting on it from a
fiber via sd_fiber_await() (synchronous-looking fiber code). sd_fiber_await()
is itself built on a "wait future" that resolves when its target resolves;
sd_future_new_wait() exposes the same primitive directly so non-fiber callers
can chain futures without involving a fiber.

Cancellation is cooperative: sd_future_cancel() invokes the future impl's 
cancel callback, which is responsible for tearing down its work and ultimately
resolving the promise with -ECANCELED. For fiber futures this is what
surfaces as the ECANCELED return from sd_fiber_yield()/sd_fiber_suspend()
mentioned above.

Fire-and-forget fibers — created by passing a NULL ret to sd_fiber_new() —
take a self-reference on their future so they outlive the caller's scope.
The self-ref is dropped when the fiber resolves. This floating mechanism
(sd_fiber_set_floating()) is restricted to fiber futures because they
uniquely guarantee resolution; allowing it for arbitrary future kinds would
risk silent leaks for kinds that may never resolve.

Note that fiber cleanup depends on the runtime operating normally. Each
fiber's _cleanup_-style cleanups live on the fiber's own stack and run
only when the fiber is resumed and allowed to unwind, which requires a
working event loop to drive it to completion. The exit event source
registered for top-level fibers ensures unwind on a normal sd_event_exit(),
but if the event loop itself terminates abnormally (e.g. an unrecoverable
allocation failure mid-dispatch) before all fibers have resolved, their
stacks never unwind and any resources they own leak.

The code lives in libsystemd as sd-future (not exported) for the following reasons:
- We may want to make this a public libsystemd API in the future
- The code can't live in src/basic as it makes heavy use of sd-event
- The code can't live in src/shared as sd-bus and sd-event make use of it

The log and log-context headers are updated with functions to allow
fibers to have their own log prefix and log context.

yuwata commented on src/systemd/sd-future.h:
```diff
@@ -0,0 +1,124 @@
+};
+
+sd_future *sd_future_ref(sd_future *f);
+sd_future *sd_future_unref(sd_future *f);
```
`sd_future *sd_future_ref(...)` -> `sd_future* sd_future_ref(...)`

yuwata commented on src/basic/fiber-def.c:
```diff
@@ -0,0 +1,25 @@
+
+DEFINE_STRING_TABLE_LOOKUP(fiber_state, FiberState);
+
+Fiber *fiber_get_current(void) {
```
`Fiber *fiber_get_...` -> `Fiber* fiber_get_...`

yuwata commented on src/basic/fiber-def.h:
```diff
@@ -0,0 +1,57 @@
+} Fiber;
+
+const char *fiber_state_to_string(FiberState s);
+FiberState fiber_state_from_string(const char *s);
```
DECLARE_STRING_TABLE_LOOKUP()

yuwata commented on src/basic/fiber-def.h:
```diff
@@ -0,0 +1,57 @@
+const char *fiber_state_to_string(FiberState s);
+FiberState fiber_state_from_string(const char *s);
+
+Fiber *fiber_get_current(void);
```
Please swap space and `*`

yuwata commented on src/libsystemd/sd-future/fiber.h:
```diff
@@ -0,0 +1,16 @@
+/* SPDX-License-Identifier: LGPL-2.1-or-later */
+#pragma once
+
+#include "sd-forward.h"
```
Please include fiber-def.h here, as Fiber object is used in the below.

daandemeyer replied:
I would prefer not to include fiber-def.h here as its more of an implementation detail. We don't want the internals of the Fiber struct to be accessible generally.

yuwata replied:
OK. Fiber is actually listed in basic-forward.h.

yuwata commented on src/libsystemd/sd-future/fiber.c:
```diff
@@ -0,0 +1,487 @@
+
+#include "alloc-util.h"
+#include "fiber-def.h"
+#include "errno-util.h"
```
Not sorted, but fiber-def.h can be dropped when fiber.h includes it.

yuwata commented on src/libsystemd/sd-future/fiber.c:
```diff
@@ -0,0 +1,487 @@
+#include "memory-util.h"
+#include "time-util.h"
+
+#if HAS_FEATURE_ADDRESS_SANITIZER
+#include <sanitizer/common_interface_defs.h>
+#endif
```
Not a matter, but maybe better to include earlier as it is a system header.

daandemeyer replied:
Not possible, HAS_FEATURE_ADDRESS_SANITIZER is defined by our own macro.h header

yuwata commented on src/libsystemd/sd-future/fiber.c:
```diff
@@ -0,0 +1,487 @@
+        stack = mmap(NULL, size,
+                     PROT_READ | PROT_WRITE,
+                     MAP_PRIVATE | MAP_ANONYMOUS | MAP_STACK,
+                     -1, 0);
```
Maybe better to add short comments for NULL, -1, and 0?

yuwata commented on src/libsystemd/sd-future/fiber.c:
```diff
@@ -0,0 +1,487 @@
+
+static thread_local ucontext_t main_context;
+
+static int fiber_allocate_stack(size_t size, void **ret) {
```
Missing assertion for `ret`.

Maybe better to refuse zero size?

Also, this does not involve fiber, maybe better to move alloc-util.c, and rename mmap_protected() or so?

yuwata commented on src/libsystemd/sd-future/fiber.c:
```diff
@@ -0,0 +1,487 @@
+
+static sd_event_source* fiber_current_event_source(Fiber *f) {
+        assert(f);
+        assert(f->state != FIBER_STATE_COMPLETED);
```
better to add `assert(f->event)`

yuwata commented on src/libsystemd/sd-future/fiber.c:
```diff
@@ -0,0 +1,487 @@
+        return sd_event_get_state(f->event) == SD_EVENT_EXITING ? f->exit_event_source : f->defer_event_source;
+}
+
+static int fiber_run(Fiber *f) {
```
Missing assertion for `f`.

yuwata commented on src/libsystemd/sd-future/fiber.c:
```diff
@@ -0,0 +1,487 @@
+        *f = (Fiber) {
+                .stack_size = ROUND_UP(buffer.rlim_cur, page_size()),
+                .state = FIBER_STATE_READY,
+                .name = strdup(name),
```
Missing assertion for `name` at the beginning. I think it is optional, so maybe `strdup_to(&f->name, name)` is better?

yuwata commented on src/libsystemd/sd-future/fiber.c:
```diff
@@ -0,0 +1,487 @@
+
+        assert(e);
+        assert(func);
+        assert(ret);
```
Missing assertion for `future`.

yuwata commented on src/libsystemd/sd-future/fiber.c:
```diff
@@ -0,0 +1,487 @@
+        return 0;
+}
+
+Fiber *fiber_free(Fiber *f) {
```
Please swap space and `*`

yuwata commented on src/libsystemd/sd-future/fiber.c:
```diff
@@ -0,0 +1,487 @@
+
+        *f = (Fiber) {
+                .stack_size = ROUND_UP(buffer.rlim_cur, page_size()),
+                .state = FIBER_STATE_READY,
```
This triggers the assertion in fiber_free() if fiber_new() failed. I think it is better to initially set FIBER_STATE_COMPLETED, then set FIBER_STATE_READY at the end of `fiber_new()`.

yuwata commented on src/libsystemd/sd-future/fiber.c:
```diff
@@ -0,0 +1,487 @@
+
+        /* We have to be able to suspend until the fiber we're waiting for finishes, and that's only
+         * possible if we're running on a fiber ourselves. */
+        assert(sd_fiber_is_running());
```
Hmm, I cannot follow the comment. Why not to simply unref() when not running?

daandemeyer replied:
Because that could end up leaking memory. We have to make sure the fiber that is cancelled can fully unwind its stack before we free it. Otherwise the necessary cleanup functions and such won't run and we'll leak memory. But as part of cancellation, a fiber can suspend again so we have to be able to suspend the current fiber to wait for it to finish.

yuwata commented on src/libsystemd/sd-future/fiber.c:
```diff
@@ -0,0 +1,487 @@
+
+        r = sd_future_cancel(f);
+        if (r < 0)
+                log_debug_errno(r, "Failed to cancel future: %m");
```
", ignoring"

yuwata commented on src/libsystemd/sd-future/fiber.c:
```diff
@@ -0,0 +1,487 @@
+
+        r = sd_fiber_await(f);
+        if (r < 0 && r != -ECANCELED)
+                log_debug_errno(r, "Failed to wait for future to finish: %m");
```
Also here.

daandemeyer committed:
sd-event: suspend instead of blocking when sd_event_run() runs on a fiber

sd_event_run() blocks the calling thread on the event loop's epoll fd
until something happens. When the caller is a fiber, that's the wrong
behaviour: blocking the thread also stalls every other fiber and the
outer event loop driving them. The most common way to hit this is a
fiber that creates its own inner event loop (e.g. a server-style fiber
that wants to dispatch its own sources independently of whatever loop
the test or supervising fiber is running on) — with the existing
implementation the inner sd_event_run() would hold the thread while the
outer scheduler should be free to advance other fibers.

Add an event_run_suspend() variant in sd-event/event-future.c that
performs the same prepare/wait/dispatch dance, but when the fast path
finds nothing ready it (a) creates an IO future watching the inner
event loop's epoll fd on the *outer* event loop, (b) optionally creates
a time future for the timeout, and (c) suspends the fiber. When either
future fires the fiber is resumed and the prepare/wait/dispatch sequence
runs once more to actually dispatch what's pending. sd_event_run()
checks sd_fiber_is_running() and delegates to this variant when on a
fiber; profile_delays accounting is intentionally skipped on that path
since the underlying prepare/wait/dispatch primitives already account
for themselves.

daandemeyer commented on src/libsystemd/sd-future/sd-future.c:
```diff
@@ -0,0 +1,503 @@
+}
+
+int sd_future_state(sd_future *f) {
+        assert(f);
```
This is still a private library, so assert is fine for now

daandemeyer committed:
sd-varlink: make sd-varlink fiber-aware

Add varlink_server_bind_fiber() and varlink_server_bind_fiber_many()
in varlink-util.{c,h} for registering a method handler that should
run on a dedicated fiber per dispatch. The fiber-bound methods live
in a separate s->fiber_methods map alongside the regular s->methods;
bind_internal()/bind_many_internal() are factored out so the regular
and fiber bind variants share their parsing/insertion code.
Registering the same method in both maps is rejected because the
dispatcher consults the regular map first and would otherwise
silently shadow the fiber binding.

varlink_dispatch_fiber() builds a VarlinkFiberData (refs to the
connection, parameters, and method name), spawns a fiber via
sd_fiber_new(), and makes the future floating so the fiber
self-manages its lifetime — neither the dispatcher nor the
connection has to track it. The fiber's priority is set to one
below the connection's quit event source so that on graceful
shutdown the fiber's exit handler fires (and runs its cleanup)
before varlink's quit_callback() closes the connection underneath
it; this is what lets a fiber-bound handler reply or flush its
sentinel on a still-open connection during shutdown.

The connection state transitions are reordered so they happen before
the fiber spawn rather than after the synchronous callback returns:
the fiber runs after dispatch has already moved past PROCESSING, which
matches the behaviour expected for a deferred reply (the fiber may
either reply immediately, or stash the connection and reply later, in
which case the post-callback logic treats it as a PENDING_METHOD).

Note that all the synchronous varlink APIs (sd_varlink_call() and friends)
already behave properly when on a fiber because they call json_stream_wait()
which calls ppoll_usec() which we already fixed to suspend when called from
a fiber.

The client/server varlink tests are migrated to fibers (threads → mock
server fibers on the same event loop) to exercise the new paths.

daandemeyer committed:
qmp-client: add fiber-aware call paths

The synchronous qmp_client_call() pumps the event loop until its reply
arrives, pinning the parsed reply on c->current so it can hand out
borrowed pointers to the caller. That model only fits one in-flight
sync call: a second qmp_client_call() on the same client clears
c->current before issuing its own send, invalidating the first
caller's borrowed pointers. On a single-threaded event loop that was
fine, but with fibers two concurrent calls on the same client can
interleave through the pump (json_stream_wait() suspends the running
fiber) and trample each other.

To fix this, make qmp_client_call() detect when it's running on a fiber 
whose event loop matches the client and transparently delegate to
qmp_client_call_suspend(), which makes use of a new QmpFuture to allow
multiple concurrent calls to qmp_client_call(). 

To make this work concurrently, we also change qmp_client_call() to 
hand out references and copies of errors so that we don't have to store
the borrowed pointers we hand out in the QmpClient struct.

daandemeyer committed:
test-qmp-client: run mock QMP servers as fibers on the shared event loop

The mock servers used to be driven out-of-band: each test created a
socketpair, forked a child, ran a hand-coded request/response script
against the raw fd, and sent SIGTERM to tear it down. That worked but
required pidref/process-util/signal plumbing in every test, two
distinct execution contexts that couldn't share state, and a JsonStream
attached to the mock side that pretended to be event-loop-driven while
actually being driven manually via blocking reads.

Now that JsonStream suspends when on a fiber, the mocks can live
inside the same process and event loop as the client. Each mock is
rewritten as an sd-fiber that runs alongside the client fiber: so the 
mock fiber yields on I/O and the event loop schedules the client in the 
meantime. Both sides progress cooperatively, no fork/SIGTERM/PID tracking, 
no manual phase tracking.

Two cleanups fall out of the rewrite:

- A QMP_TEST(name, mock_fn) { ... } macro encapsulates the per-test
  scaffolding (event loop, socketpair, mock fiber spawn, exit-on-idle
  shim) and injects an already-connected QmpClient *client into the
  test body. Each test now reads as a flat sequence of
  qmp_client_call() invocations against that client.

- Repeated mock command/reply scripting is factored into
  mock_qmp_expect(), mock_qmp_reply(), mock_qmp_expect_and_reply(),
  mock_qmp_handshake(), and mock_qmp_query_status_running(). The
  greeting JSON is built with sd_json_buildo() instead of being parsed
  from a literal.

The file shrinks from 756 to 494 lines, mostly through deletions.

daandemeyer commented on src/basic/log.c:
```diff
@@ -462,8 +482,8 @@ static int write_to_console(
                 iovec[n++] = IOVEC_MAKE_STRING(on);
-        if (log_prefix) {
-                iovec[n++] = IOVEC_MAKE_STRING(log_prefix);
+        if (log_prefix()) {
```
overkill

daandemeyer commented on src/libsystemd/sd-bus/sd-bus.c:
```diff
@@ -222,6 +223,12 @@ static sd_bus* bus_free(sd_bus *b) {
 
+        /* Outstanding fiber handlers pin the bus via their BusFiberData ref, so by the time refcount
+         * reaches zero and bus_free() runs, every fiber has already resolved and removed itself from
+         * this set. */
```
assert is not compiled out

daandemeyer commented on src/libsystemd/sd-future/sd-future.c:
```diff
@@ -0,0 +1,268 @@
+        /* By convention the first field of any impl struct is `sd_promise *promise` so handlers that
+         * receive the impl pointer can resolve the future without a separate lookup. Stamp the
+         * back-pointer here so the caller doesn't have to. */
+        *(sd_promise **) impl = &f->promise;
```
This doesn't really help with anything

daandemeyer commented on src/libsystemd/sd-bus/bus-objects.c:
```diff
@@ -407,6 +471,50 @@ static int method_callbacks_run(
 
                 slot = container_of(c->parent, sd_bus_slot, node_vtable);
 
+                if (FLAGS_SET(c->vtable->flags, SD_BUS_VTABLE_METHOD_FIBER)) {
```
assertion is fine

daandemeyer commented on src/libsystemd/sd-bus/sd-bus.c:
```diff
@@ -3177,7 +3203,14 @@ static int process_closing(sd_bus *bus, sd_bus_message **ret) {
+         * before tearing down the rest of the bus. bus_enter_closing() scheduled the cancel; each
+         * fiber resolves asynchronously and bus_fiber_resolved() removes it from the set. Returning
+         * 1 here keeps the bus in CLOSING state so the event loop drives the fibers to completion. */
+        if (!set_isempty(bus->fiber_futures))
```
This is fine as is

daandemeyer commented on src/shared/qmp-client.h:
```diff
@@ -72,16 +75,47 @@ int qmp_client_call(
+
+/* Borrowed-lifetime extraction of the reply from a resolved qmp_client_call_future(). result
+ * and error_desc remain valid until f is freed. */
+int future_get_qmp_reply(
```
This is intended to match other future functions

daandemeyer committed:
sd-future: make src/basic blocking helpers fiber-aware

Some helpers in src/basic — ppoll_usec_full() (used by fd_wait_for_event()),
loop_read(), loop_read_exact(), loop_write_full() and
pidref_wait_for_terminate_full() — block the calling thread. That's the
right behaviour outside a fiber but not inside one, where blocking the
thread also stalls every other fiber running on the same event loop.
Rewriting every caller to pick a fiber or non-fiber variant explicitly
would be a lot of churn and would split otherwise-shared code paths in
two.

Instead, the helpers detect at runtime whether they're running on a fiber
and dispatch to a suspending variant when they are. FiberOps in
fiber-ops.h holds five function pointers (ppoll, read, write, timeout,
cancel_wait_unref); a fiber_ops global constant is populated whenever we 
enter a fiber with functions that delegate to suspending variants of common
syscalls. With this approach, the variants themselves stay in libsystemd
which is required because they make use of sd-event.

- loop_read()/loop_read_exact() take the fiber read hook on a fiber
  unless the caller asked for a non-blocking attempt (do_poll=false) and
  the fd is already non-blocking — in that case we fall through to read()
  to preserve the existing return-EAGAIN-immediately semantic. The hook
  itself suspends on EAGAIN until data is available, so neither the
  do_poll knob nor the explicit fd_wait_for_event() retry loop are
  needed on the fiber path.

- loop_write_full() likewise takes the fiber write hook on a fiber,
  except when timeout=0 with an already-non-blocking fd (preserving the
  fast-return-EAGAIN semantic). The fiber path runs inside a
  FIBER_OPS_TIMEOUT() scope so the caller's timeout is honoured via a
  deadline future, mirroring SD_FIBER_TIMEOUT() but reachable from
  src/basic without pulling in sd-future.h.

- pidref_wait_for_terminate_full() polls the pidfd via fd_wait_for_event()
  before each waitid() when either a finite timeout is set or we're on a
  fiber, and requires pidref->fd >= 0 in those cases (returning
  -ENOMEDIUM otherwise — extending the rule that already applied to
  finite timeouts). The poll suspends the fiber via the ppoll hook above;
  the subsequent waitid() doesn't block because the pidfd is already
  signalled.

daandemeyer committed:
sd-future: add fiber-aware non-blocking I/O wrappers

Add a family of sd_fiber_*() I/O wrappers that, when called from a
fiber, behave like blocking I/O from the caller's perspective but
yield to the event loop instead of blocking the thread:

  sd_fiber_read / sd_fiber_write
  sd_fiber_readv / sd_fiber_writev
  sd_fiber_recv / sd_fiber_send
  sd_fiber_connect
  sd_fiber_recvmsg / sd_fiber_sendmsg
  sd_fiber_recvfrom / sd_fiber_sendto
  sd_fiber_accept
  sd_fiber_ppoll

Most of them share a single helper, fiber_io_operation(), which when
invoked outside a fiber falls through to the underlying syscall
directly, preserving the regular blocking behaviour. Inside a fiber
the helper flips the fd to non-blocking (restoring its original mode
on return), tries the syscall once on the fast path, and on EAGAIN/
EWOULDBLOCK creates an sd-event-backed IO future via future_new_io(),
suspends the fiber, and retries the syscall once the event source
fires.

future_new_io() itself is added to sd-event/event-future.{c,h} as a
new IoFuture kind. It wraps sd_event_add_io() into an sd_future:
oneshot enable, EPOLLERR translated via SO_ERROR (suppressed for
non-sockets), and the fd duplicated with F_DUPFD_CLOEXEC to avoid
EEXIST when multiple sources watch the same descriptor.

Together these let fiber-using code write straight-line socket and
pipe I/O without bundling state into callbacks.

daandemeyer commented on src/libsystemd/sd-future/fiber-io.c:
```diff
@@ -0,0 +1,423 @@
+#include <sys/epoll.h>          /* IWYU pragma: keep */
+#include <sys/socket.h>
+#include <sys/uio.h>
+#include <time.h>
```
is used, CLOCK_MONOTONIC

daandemeyer commented on src/libsystemd/sd-future/fiber.c:
```diff
@@ -0,0 +1,620 @@
+        case FIBER_STATE_READY:
+                log_debug("Fiber yielded execution");
+
+                r = sd_event_source_set_enabled(fiber_current_event_source(f), SD_EVENT_ONESHOT);
```
error left to the caller, amended commit message mentioning that we need the event loop to work properly to guarantee cleanup

daandemeyer commented on src/libsystemd/sd-future/fiber.c:
```diff
@@ -0,0 +1,620 @@
+        if (IN_SET(f->state, FIBER_STATE_COMPLETED, FIBER_STATE_CANCELLED))
+                return 0;
+
+        f->state = FIBER_STATE_CANCELLED;
```
Same thing as below, we depend on the event loop working properly

daandemeyer commented on src/libsystemd/sd-future/fiber-io.c:
```diff
@@ -0,0 +1,422 @@
+                reset_fd = fd;
+
+        ssize_t n = func(fd, args);
+        if (n >= 0 || !ERRNO_IS_NEG_TRANSIENT(n))
```
Shouldn't get EINTR here

poettering commented on src/basic/fiber-def.h:
```diff
@@ -0,0 +1,64 @@
+        sd_promise *promise;            /* must be first: set by sd_future_new() via the impl convention */
+
+        void *stack;
+        size_t stack_size;
```
struct iovec?

poettering commented on src/systemd/sd-future.h:
```diff
@@ -0,0 +1,100 @@
+        void* (*free)(void *userdata);
+        int (*cancel)(void *userdata);
+        int (*set_priority)(void *userdata, int64_t priority);
+};
```
hmm, an open coded struct? why?

poettering commented on src/systemd/sd-future.h:
```diff
@@ -0,0 +1,100 @@
+enum {
+        SD_FUTURE_PENDING,
+        SD_FUTURE_RESOLVED
+};
```
in newer code please use named enums, but declare them with:

```
__extension__ typedef enum _SD_ENUM_TYPE_S64(…)
```

so that they are abi stable even if extended.

poettering commented on src/libsystemd/sd-future/sd-future.c:
```diff
@@ -0,0 +1,268 @@
+        void *userdata;
+
+        const sd_future_ops *ops;
+        void *impl;
```
hmm, is there a comment anywhere what `impl` is? please add it, probably here

poettering replied:
rename this into "private" I think. and please add a pair of comments here that make clear which fields are for the consumer and which for the implementor. i.e.:

```c
/* Callback + pointer provided by the consumer */
sd_future_func_t callback;
void *userdata;

/* Callbacks + pointer provided by the implementor */
sd_future_free_func_t free_callback;
sd_future_cancel_func_t cancel_callback;
sd_future_set_priority_func_t set_priority_callback;
void *private;
```

or so

poettering commented on src/systemd/sd-future.h:
```diff
@@ -0,0 +1,100 @@
+typedef struct sd_promise sd_promise;
+typedef int (*sd_future_func_t)(sd_future *f);
+typedef int (*sd_fiber_func_t)(void *userdata);
+typedef void (*sd_fiber_destroy_t)(void *userdata);
```
maybe do:

```c
typedef _sd_destroy_t sd_fiber_destroy_t;
```

poettering commented on src/systemd/sd-future.h:
```diff
@@ -0,0 +1,100 @@
+sd_future* sd_future_cancel_wait_unref(sd_future *f);
+_SD_DEFINE_POINTER_CLEANUP_FUNC(sd_future, sd_future_cancel_wait_unref);
+void sd_future_cancel_wait_unref_array_clear(sd_future **array, size_t n);
+void sd_future_cancel_wait_unref_array(sd_future **array, size_t n);
```
i'd always declare the params here ars `sd_future *array[]` rather than `sd_future **array`, as it makes very clear it's an array i think

poettering commented on src/systemd/sd-future.h:
```diff
@@ -0,0 +1,100 @@
+int sd_future_state(sd_future *f);
+int sd_future_result(sd_future *f);
+void* sd_future_get_userdata(sd_future *f);
+void* sd_future_get_impl(sd_future *f);
```
hmm, what's userdata? what's impl? this is a bit opaque... i guess it is about ownership (i.e. owned by user of a future vs. implementationm of a future), but this is not visible here, and really deserves some explanation.

i wish there were docs with this, i.e. a man page that explains what this is.

also, not sure i like the abbreviation "impl"...

poettering commented on src/systemd/sd-future.h:
```diff
@@ -0,0 +1,100 @@
+
+int sd_fiber_is_running(void);
+sd_future* sd_fiber_get_current(void);
+const char* sd_fiber_get_name(sd_future *f);
```
is "name" the right word? shouldn't this be "description"? i.e. to match our wording in sd_event_source_set_description()?

poettering commented on src/libsystemd/sd-event/event-future.c:
```diff
@@ -0,0 +1,127 @@
+
+static int time_future_cancel(void *impl) {
+        TimeFuture *f = ASSERT_PTR(impl);
+        int r = sd_event_source_set_enabled(f->source, SD_EVENT_OFF);
```
we generall want `int r` to be defined on a separate line early on

poettering commented on src/libsystemd/sd-event/event-future.c:
```diff
@@ -0,0 +1,127 @@
+        .free = time_future_free,
+        .cancel = time_future_cancel,
+        .set_priority = time_future_set_priority,
+};
```
i really don't like that this is open coded. it feels like we might need more callbacks sooner rather than later, and then this here blows up in our face, given it's public facing.

just do sd_future_set_free_callback() and so on?

poettering commented on meson.build:
```diff
@@ -1019,6 +1019,13 @@ librt = cc.find_library('rt')
+else
+        libucontext = dependency('libucontext', required : get_option('libucontext'))
+endif
+conf.set10('HAVE_LIBUCONTEXT', libucontext.found())
```
oh my god oh my god, @daandemeyer of all people adds a non-dlopen dep to systemd, kinda

poettering commented on src/libsystemd/sd-event/event-future.c:
```diff
@@ -0,0 +1,127 @@
+#include "errno-util.h"
+#include "event-future.h"
+
+typedef struct TimeFuture {
```
couldn't the time future stuff also be added in a separate commit?

poettering commented on src/libsystemd/sd-future/fiber.c:
```diff
@@ -0,0 +1,616 @@
+        if (!fiber)
+                return -ENOMEM;
+
+        struct rlimit rl = { .rlim_cur = FIBER_DEFAULT_STACK_SIZE };
```
frankly, i'd just do:

```c
struct rlimit rl = { RLIM_INFINITY, RLIM_INFINITY },
```

and then let the check below map the fallback case to FIBER_DEFAULT_STACK_SIZE once

poettering commented on src/libsystemd/sd-future/fiber.c:
```diff
@@ -0,0 +1,616 @@
+        stack = mmap(/* addr= */ NULL, size,
+                     PROT_READ | PROT_WRITE,
+                     MAP_PRIVATE | MAP_ANONYMOUS | MAP_STACK,
+                     /* fd= */ -1, /* offset= */ 0);
```
-EBADF?

poettering commented on src/libsystemd/sd-future/fiber.c:
```diff
@@ -0,0 +1,616 @@
+                (void) munmap(stack, size);
+                return r;
+        }
+
```
iirc there are still relevant archs where the stack grows up, not down.

also, if you realize the guard page here, maybe add the extra page to the size calculations here too, rather than let the caller do it, so that the guard page magic is entirely hidden in this function here

poettering commented on src/libsystemd/sd-future/fiber.c:
```diff
@@ -0,0 +1,616 @@
+        /* __register_atfork() either returns 0 or -ENOMEM, in its glibc implementation. Since it's
+         * only half-documented (glibc doesn't document it but LSB does — though only superficially)
+         * we'll check for errors only in the most generic fashion possible. */
+        atfork_ret = pthread_atfork(NULL, NULL, reset_current_fiber);
```
parameter comments plz for the first two args

poettering commented on src/libsystemd/sd-future/sd-future.c:
```diff
@@ -0,0 +1,268 @@
+struct sd_promise {};
+
+struct sd_future {
+        sd_promise promise;
```
i think i would prefer if the promise stuff would be done with a macro or inline func that casts. i find this embedded embedded struct a bit opaque, since it suggests promise was a somehow a subconcept of a future, but it's the same thing, just seen from a different pov.

Or alternatively, maybe not do this at all, it just complicates understanding? it appears to me that futures are not going to be implemented by millions of people, but mostly consumed by them, hence it seems unnecessary try to lock them down in some way, if only very few people might run into this.

poettering commented on src/libsystemd/sd-future/sd-future.c:
```diff
@@ -0,0 +1,268 @@
+        /* By convention the first field of any impl struct is `sd_promise *promise` so handlers that
+         * receive the impl pointer can resolve the future without a separate lookup. Stamp the
+         * back-pointer here so the caller doesn't have to. */
+        *(sd_promise **) impl = &f->promise;
```
uff. not sure i like this.

how is this just a "convention" if you always initialize stuff for people? 

either the "impl" pointer is abstract or it isn't, but this here seems really weird.

poettering commented on src/libsystemd/sd-future/sd-future.c:
```diff
@@ -0,0 +1,268 @@
+
+int sd_future_result(sd_future *f) {
+        assert(f);
+        assert(f->state == SD_FUTURE_RESOLVED);
```
maybe better just return -ENODATA here?

poettering commented on src/libsystemd/sd-future/sd-future.c:
```diff
@@ -0,0 +1,268 @@
+        sd_future *f = container_of(p, sd_future, promise);
+        int r = 0;
+
+        assert(f);
```
our sd-xyz apis typically do not hard assert on things, but use assert_return() and soft-assert, because we do not want to require people to get things right as hard as we require it from ourselves.

(here and everywhere)

poettering commented on src/libsystemd/sd-future/fiber.c:
```diff
@@ -0,0 +1,616 @@
+#include <sanitizer/common_interface_defs.h>
+#endif
+
+#define FIBER_DEFAULT_STACK_SIZE UINT64_C(8388608)
```
write as `(8U * U64_MB)` please

daandemeyer commented on src/basic/fiber-ops.c:
```diff
@@ -0,0 +1,50 @@
+        return fiber_ops->timeout(timeout);
+}
+
+sd_future* fiber_ops_cancel_wait_unref(sd_future *f) {
```
This should never be called outside of a fiber context, so the assert is correct.

daandemeyer commented on src/libsystemd/sd-event/event-future.h:
```diff
@@ -0,0 +1,7 @@
+/* SPDX-License-Identifier: LGPL-2.1-or-later */
+#pragma once
+
+#include "sd-forward.h"
+
+int future_new_time(sd_event *e, clockid_t clock, uint64_t usec, uint64_t accuracy, int result, sd_future **ret);
```
sd-forward.h provides these includes

daandemeyer commented on src/shared/qmp-client.c:
```diff
@@ -810,21 +819,174 @@ int qmp_client_invoke(
+
+        QmpFuture *qf = sd_future_get_private(f);
+
+        r = qmp_client_send(c, command, args, qmp_future_callback, f, &qf->slot);
```
The callback is canceled if the future is freed

daandemeyer committed:
sd-bus: make sd-bus fiber-aware

Two changes to teach sd-bus how to behave when called from a fiber, in
order of increasing depth:

2. sd_bus_call() now redirects to a new bus_call_suspend() helper when
   the caller is a fiber whose event loop is the same one the bus is
   attached to. The plain bus_poll() path serializes all bus traffic on
   the slot's reply (only one method call can be in flight per
   sd_bus*), which would defeat the point of running multiple fibers
   against one bus. bus_call_suspend() builds on the async sd-bus API:
   it wraps the call in a new BusFuture (sd-bus/bus-future.{c,h}) that
   resolves when the reply or method-error arrives, lets the fiber
   await that future, and surfaces the reply to the caller via
   future_get_bus_reply(). Because the futures live on the event loop
   rather than a per-bus slot, multiple fibers can drive concurrent
   method calls against the same bus.

3. A new private SD_BUS_VTABLE_METHOD_FIBER flag dispatches a vtable
   method handler on its own fiber, so handlers are free to use
   sd_bus_call() against the same bus, sd_fiber_sleep(), loop_read(),
   etc. without stalling the event loop for other connections or
   handlers. The flag stays out of sd-bus-vtable.h (its bit value is
   reserved there to prevent collisions) — the fiber runtime is a
   systemd-internal implementation detail.

Lifecycle of fiber-dispatched handlers is tracked on the bus itself: a
new bus->fiber_futures set holds a ref to each in-flight handler.
bus_enter_closing() cancels every entry and process_closing() returns
with the bus still in CLOSING state until the set drains, so we can be
sure no fiber handler outlives the bus. bus_fiber_resolved() removes
the entry on completion. bus_free()'s assert(set_isempty()) makes the
invariant load-bearing.

Note that plain sd_bus_call() already works correctly on a fiber as it
calls ppoll_usec() which has already been modified to suspend when 
running on a fiber.

To exercise these changes the existing thread-based client/server
sd-bus tests (test-bus-chat, test-bus-objects, test-bus-peersockaddr,
test-bus-server, test-bus-watch-bind) are migrated to fibers, and a
new test-bus-fiber is added that covers SD_BUS_VTABLE_METHOD_FIBER —
including handlers that issue nested sd_bus_call() on the same bus, the
cancel-on-close path, and concurrent dispatches across multiple fibers.

daandemeyer commented on src/libsystemd/sd-varlink/sd-varlink.c:
```diff
@@ -956,6 +958,178 @@ static int generic_method_get_interface_description(
+                         * outside the dispatch stack frame (e.g. called from varlink_fiber_entry after
+                         * the fiber returned), so we go straight to IDLE_SERVER ourselves. PROCESSING_*
+                         * means we're inside varlink_dispatch_method(), which will transition us. */
+                        if (IN_SET(v->state, VARLINK_PENDING_METHOD, VARLINK_PENDING_METHOD_MORE)) {
```
Preserving MORE semantics there would be a bug: the fiber is gone, and the client must see this reply as final.

daandemeyer commented on src/libsystemd/sd-event/event-future.c:
```diff
@@ -0,0 +1,247 @@
+
+        IoFuture *iof = sd_future_get_private(f);
+
+        /* Duplicate fd to avoid EEXIST from epoll when adding the same fd multiple times */
+        _cleanup_close_ int fd_copy = fcntl(fd, F_DUPFD_CLOEXEC, 3);
```
Bogus

daandemeyer commented on src/libsystemd/sd-future/fiber.c:
AFAICT not done by other fiber implementations either.

poettering commented on src/basic/log.c:
```diff
@@ -87,6 +87,14 @@ bool _log_message_dummy = false; /* Always false */
+
+        const char *prev = log_prefix;
+        log_prefix = *prefix;
+        *prefix = prev;
```
SWAP_TWO()

poettering commented on src/libsystemd/sd-future/fiber.c:
```diff
@@ -0,0 +1,947 @@
+ * stack and can be at any address relative to the caller. Disable fortify here so siglongjmp resolves
+ * to the plain glibc entry point. */
+#undef _FORTIFY_SOURCE
+#define _FORTIFY_SOURCE 0
```
shouldn't this just be disabled around siglongjmp()

poettering commented on src/libsystemd/sd-future/fiber.c:
```diff
@@ -0,0 +1,947 @@
+
+#if HAS_FEATURE_ADDRESS_SANITIZER
+#include <sanitizer/common_interface_defs.h>
+#endif
```
why down here

daandemeyer replied:
The macro comes from macro.h

poettering commented on src/libsystemd/sd-future/fiber.c:
maybe abstract this in some header file, i.e. add STACK_GROWS_UP macro?

poettering commented on src/libsystemd/sd-future/fiber.c:
no {} around single line if blocks

daandemeyer commented on src/libsystemd/sd-future/fiber.c:
It deals with both positive and negative

daandemeyer commented on src/shared/main-func.c:
```diff
@@ -34,3 +39,86 @@ int exit_failure_if_negative(int result) {
+        if (r < 0)
+                return log_error_errno(r, "Failed to install SIGINT handler: %m");
+
+        r = sd_event_add_signal(event, &sigterm_source, SIGTERM | SD_EVENT_SIGNAL_PROCMASK,
```
Fibers are cleaned up during event loop exit

poettering commented on src/basic/log-context.c:
```diff
@@ -177,6 +177,19 @@ size_t log_context_num_fields(void) {
+
+        LogContext *prev_log_context = _log_context;
+        _log_context = *log_context;
+        *log_context = prev_log_context;
```
SWAP_TWO?

poettering commented on src/basic/log-context.c:
```diff
@@ -177,6 +177,19 @@ size_t log_context_num_fields(void) {
+
+        size_t prev_log_context_num_fields = _log_context_num_fields;
+        _log_context_num_fields = *num_fields;
+        *num_fields = prev_log_context_num_fields;
```
ditto

poettering commented on src/basic/log-context.h:
```diff
@@ -66,6 +66,10 @@ size_t log_context_num_contexts(void);
 /* Returns the number of fields in all attached log contexts. */
 size_t log_context_num_fields(void);
 
+/* Atomically swap the thread-local log context list and field count with the values pointed to by *list and
```
"Atomically" is a bit of an overstatement

poettering commented on src/basic/log.h:
```diff
@@ -380,6 +380,10 @@ int log_syntax_parse_error_internal(
 
 const char* _log_set_prefix(const char *prefix, bool force);
+
+/* Atomically swap the thread-local log prefix with the value pointed to by *prefix. Used by sd-fiber to
```
here too

usually people like me think atomic cpu ops when they hear "atomic", but this is hardly it

poettering commented on src/libsystemd/sd-event/event-future.c:
```diff
@@ -0,0 +1,116 @@
+        return new0(TimeFuture, 1);
+}
+
+static void time_future_free(sd_future *f) {
```
shouldn't this func return `(sd_future*) NULL`?

daandemeyer replied:
Well it doesn't free the entire future, felt a bit weird to make it return sd_future

poettering commented on src/libsystemd/sd-event/event-future.c:
```diff
@@ -0,0 +1,116 @@
+
+static int time_future_cancel(sd_future *f) {
+        TimeFuture *tf = sd_future_get_private(ASSERT_PTR(f));
+        int r = sd_event_source_set_enabled(tf->source, SD_EVENT_OFF);
```
coding style: we usually put "int r" on a separate definition up top

poettering commented on src/libsystemd/sd-event/event-future.c:
```diff
@@ -0,0 +1,116 @@
+}
+
+static int time_future_set_priority(sd_future *f, int64_t priority) {
+        TimeFuture *tf = sd_future_get_private(ASSERT_PTR(f));
```
ASSERT_PTR() also around sd_future_get_private() given you are going to dereference below

poettering commented on src/libsystemd/sd-event/event-future.c:
```diff
@@ -0,0 +1,116 @@
+
+static int time_handler(sd_event_source *s, usec_t usec, void *userdata) {
+        sd_future *f = ASSERT_PTR(userdata);
+        TimeFuture *tf = sd_future_get_private(f);
```
as above

poettering commented on src/libsystemd/sd-event/event-future.c:
```diff
@@ -0,0 +1,116 @@
+        if (r < 0)
+                return r;
+
+        TimeFuture *tf = sd_future_get_private(f);
```
as above (here and everywhere)

poettering commented on src/libsystemd/sd-event/event-future.c:
```diff
@@ -0,0 +1,116 @@
+        if (r < 0)
+                return r;
+
+        if (sd_fiber_is_running()) {
```
hmm, what's the explanation for not considering the original thread a process starts with already to be a fiber too just a more important one?

daandemeyer replied:
Well, if we considered it a fiber, where would we swap to if it called a suspending function? There's nowhere it can swap to, so we have to consider it special so it does blocking function calls rather than non blocking function calls.

But I have thought about this, and the way I'm going to handle this is to add `DEFINE_MAIN_FUNCTION_FIBER()` which runs `run()` in a fiber, so that you're in fiber land from the very beginning.

poettering commented on src/libsystemd/sd-future/fiber.c:
```diff
@@ -0,0 +1,951 @@
+        sigjmp_buf context;             /* Where to jump to when entering or resuming the fiber. */
+        sigjmp_buf resume_context;      /* Where to jump back to when the fiber yields or completes. */
+
+        /* Caller's stack range, recorded by fiber_run() on each entry so the fiber's siglongjmp back
```
() after siglongjmp

poettering commented on src/libsystemd/sd-future/fiber.c:
```diff
@@ -0,0 +1,951 @@
+        current_fiber = f;
+}
+
+static int fiber_allocate_stack(size_t size, void **ret) {
```
please add a comment here clarifying that the effective stack size is one page less than this size, because the guard page is subtracted

poettering commented on src/libsystemd/sd-future/fiber.c:
what's this good for?

poettering commented on src/libsystemd/sd-future/fiber.c:
suffix () to siglongjmp

poettering commented on src/libsystemd/sd-future/fiber.c:
assert(f);

poettering commented on src/libsystemd/sd-future/fiber.c:
comment out of date?

poettering commented on src/libsystemd/sd-future/fiber.c:
prefix 2nd param with `/* val= */`

poettering commented on src/libsystemd/sd-future/fiber.c:
`/* val= */`

poettering commented on src/libsystemd/sd-future/fiber.c:
`/* argc= */`

poettering commented on src/libsystemd/sd-future/fiber.c:
param comment on 2nd param (here and everywhere)

poettering commented on src/libsystemd/sd-future/fiber.c:
```
sigprocmask_many(SIG_BLOCK, &osigs, SIGURG);
```

poettering replied:
or just use BLOCK_SIGNALS()?

poettering commented on src/libsystemd/sd-future/fiber.c:
why do we need this alternative implementation again? just to avoid a dep on libucontext if musl is used? why do we care about that? musl is always gonna be half-assed, not sure it's worth the effort?

daandemeyer replied:
I think this is overall less pain than the libucontext dependency. Third party dependencies always carry with them some bureaucracy. Given doing it ourselves is easy enough, I'd prefer to do it like that rather than introducing the dep, even if it's just for musl. Yes this is ugly as it messes with signal handlers and such from a library, but we get none of that on glibc, and the musl users will have to deal with the fallout of musl not supporting the ucontext.h interface.

Or they patch downstream to link with libucontext after all and we'll automatically pick it up via the weak symbol anyway. But this approach keeps things working, including in our own musl CI without having to bother messing around with extra deps.

daandemeyer replied:
Yes musl is half assed, but if we start using fibers everywhere and it doesn't work on musl then it's not half assed, it's just completely broken. Hence this alternative implementation, which I would still call half assed as it messes with signal handlers from a library but since it's musl I don't really care. On glibc we get the nicer behavior via `makecontext()` and `swapcontext()`

bluca replied:
> why do we need this alternative implementation again? just to avoid a dep on libucontext if musl is used? why do we care about that? musl is always gonna be half-assed, not sure it's worth the effort?

I agree with Lennart, I too would prefer to avoid open coding this, and instead just make libucontext a requirement for musl builds. It's just one CI job where to add it and it's widely availiable, so shouldn't be a big issue?

poettering commented on src/libsystemd/sd-future/fiber.c:
maybe add a `// FIXME: delete when baseline above 6.13`

brauner commented on src/libsystemd/sd-future/fiber.c:
You should note that atfork handlers aren't run for clone()/clone3()/raw_clone() and posix_spawn().
